src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/IQueueItem.cs

```diff
@@ -17,33 +17,41 @@ public interface IQueueItem<TRequestContext>
     /// <summary>
     /// Executes the work specified by this queue item.
     /// </summary>
-    /// <param name="requestContext">the context created by <see cref="CreateRequestContextAsync(CancellationToken)"/></param>
+    /// <param name="requestContext">the context created by <see cref="CreateRequestContextAsync(IMethodHandler, CancellationToken)"/></param>
     /// <param name="cancellationToken" />
     /// <returns>A <see cref="Task "/> which completes when the request has finished.</returns>
-    Task StartRequestAsync(TRequestContext? requestContext, CancellationToken cancellationToken);
+    Task StartRequestAsync(IMethodHandler handler, TRequestContext requestContext, CancellationToken cancellationToken);
 
     /// <summary>
     /// Creates the context that is sent to the handler for this queue item.
     /// Note - this method is always called serially inside the queue before
-    /// running the actual request in <see cref="StartRequestAsync(TRequestContext?, CancellationToken)"/>
+    /// running the actual request in <see cref="StartRequestAsync(IMethodHandler, TRequestContext, CancellationToken)"/>
     /// Throwing in this method will cause the server to shutdown.
     /// </summary>
-    Task<TRequestContext?> CreateRequestContextAsync(CancellationToken cancellationToken);
-
-    ILspServices LspServices { get; }
+    Task<TRequestContext> CreateRequestContextAsync(IMethodHandler handler, CancellationToken cancellationToken);
 
     /// <summary>
-    /// Indicates that this request may mutate the server state, so that the queue may handle its execution appropriatly.
+    /// Gets the identifier of the document from the request, if the request provides one.
     /// </summary>
-    bool MutatesServerState { get; }
+    /// <typeparam name="TTextDocumentIdentifier">The type of the text document identifier.</typeparam>
+    /// <param name="handler">The handler for the request.</param>
+    /// <returns>The text document identifier.</returns>
+    TTextDocumentIdentifier GetTextDocumentIdentifier<TTextDocumentIdentifier>(IMethodHandler handler);
```

Comment:
I think this needs to be explicitly nullable right?

Reply:
Yes, done.